Bugfix/mtm 59757/user scope in microservice sdk draft (#3107)
SoftwareAG/c8y-docs: content/microservice-sdk/java-bundle/developing-microservice.md

---

> The {{< product-c8y-iot >}} Microservice SDK is an open source toolkit for building microservices using the popular [Spring Boot](https://spring.io/projects/spring-boot) framework and the {{< product-c8y-iot >}} Platform API. It accelerates development by offering preconfigured annotations, built-in services, and a Maven plugin for creating Docker containers and Cumulocity applications. The SDK source code is available on GitHub in the

Reviewer: ```suggestion
The {{< product-c8y-iot >}} Microservice SDK is an open source toolkit for building microservices using the popular [Spring Boot](https://spring.io/projects/spring-boot) framework and the {{< product-c8y-iot >}} Platform API. It accelerates development by offering preconfigured annotations, built-in services, and a Maven plugin for creating Docker containers and {{< product-c8y-iot >}} applications. The SDK source code is available on GitHub in the
```

> The {{< product-c8y-iot >}} Microservice SDK is an open source toolkit for building microservices using the popular [Spring Boot](https://spring.io/projects/spring-boot) framework and the {{< product-c8y-iot >}} Platform API. It accelerates development by offering preconfigured annotations, built-in services, and a Maven plugin for creating Docker containers and {{< product-c8y-iot >}} applications. The SDK source code is available on GitHub in the

Reviewer: @Robert-Pretz Does this refer to the Platform API desribed here? https://cumulocity.com/api/core/#tag/Platform-API Then we might add the link. In general I see a mix of "Platform API" (capital letter) and "platform APIs". Does the latter refer in general to all Cumulocity APIs? This is not clear to me. If it refers to the one "Platform API" it should be spelled with capital letter.

Reviewer 2: I changed it to platform APIs and linked the docs. I think the cases are in general not a problem, upper case is simply the name and lower case the general term, but there is little room for confusion (even though there is the Platform REST API resource).

Reviewer 2: I changed it to platform APIs and linked the docs. I think the cases are in general not a problem, upper case is simply the name and lower case the general term, but there is little room for confusion (even though there is the Platform REST API resource).

[cumulocity-clients-java](https://github.com/Cumulocity-IoT/cumulocity-clients-java) repository.

In this section, you will learn how to:

> * Use SDK annotations to simplify setup.

Reviewer: ```suggestion
* Use SDK annotations to simplify the setup.
```

* Access platform APIs via dependency injection.
* Authenticate towards the platform APIs.
* Secure access to your microservice APIs.
* Handle subscriptions and multi-tenant access.
* Configure your microservice.
* Upload and run your microservice.
* Manage and monitor your microservice.
* Set up external or legacy deployments.

### Using annotations {#annotations}

> As shown in the [in the "Hello world" tutorial](/microservice-sdk/java/#create-a-java-application), the easiest way to enable default microservice behavior is to annotate your main class with `@MicroserviceApplication`. This composite annotation includes:

Reviewer: ```suggestion
As shown in the ["Hello world" tutorial](/microservice-sdk/java/#create-a-java-application), the easiest way to enable default microservice behavior is to annotate your main class with `@MicroserviceApplication`. This composite annotation includes:
```

| -------------------------------------- | ------------------------------------------------------------------------------------------------ |
| @SpringBootApplication                 | Enables Spring Boot’s auto-configuration                                                         |
| @EnableContextSupport                  | Allows use of @UserScope and @TenantScope for method-level context switching                     |
| @EnableHealthIndicator                 | Exposes a standard health endpoint for platform monitoring                                       |
| @EnableMicroserviceSecurity            | Enables security by verifying users and roles against the platform                               |
| @EnableMicroserviceSubscription        | Manages subscriptions, metadata updates, and listens to tenant subscription changes              |
| @EnableMicroservicePlatformInternalApi | Injects platform API services into the Spring context                                            |
| @EnableTenantOptionSettings            | Allows configuration through tenant options and supports overriding default properties via files |

> ### Accessing platform APIs {#platform-api}

Reviewer: ```suggestion
### Accessing platform APIs {#acessing-platform-api}
```

* SystemOption - SystemOptionApi
* Token - TokenApi
* Notification - NotificationSubscriptionApi

Each API supports standard "CRUD" operations (create, read, update, delete). For example, the `AlarmApi` interface provides:

```java
// Create
AlarmRepresentation create(AlarmRepresentation alarm) throws SDKException;
Future createAsync(AlarmRepresentation alarm) throws SDKException;
// Read
AlarmRepresentation getAlarm(GId gid) throws SDKException;
AlarmCollection getAlarms() throws SDKException;
AlarmCollection getAlarmsByFilter(AlarmFilter filter) throws SDKException;
// Update
AlarmRepresentation updateAlarm(AlarmRepresentation alarm) throws SDKException;
// Delete
void deleteAlarmsByFilter(AlarmFilter filter) throws IllegalArgumentException, SDKException;
```

This is an example of creating an alarm:

```java
@Autowired
private AlarmApi alarms;

public AlarmRepresentation addHelloAlarm () {
    AlarmRepresentation alarm = new AlarmRepresentation();
    alarm.setSeverity("CRITICAL");
    alarm.setStatus("Hello");

    return alarms.create(alarm);
}
```

More details on using the APIs are available in the [Client library](/microservice-sdk/java/#client-library) section.

> ### Authenticating and authorizing towards the platform {#using-platform-api}

Reviewer: ```suggestion
### Authenticating and authorizing towards the platform {#authenticating-and-authorizing-towards-the-platform}
```

API requests in {{< product-c8y-iot >}} microservices can run in two authentication scopes:
 * Tenant scope – uses the service user's credentials.
 * User scope – uses the credentials of the authenticated user who triggered the request.

> Each microservice has a service user whose roles are defined in the `cumulocity.json` manifest. By default, the Platform API services provided by the Microservice SDK operate in the tenant scope. The following example uses the predefined eventApi bean in tenant scope:

Reviewer: ```suggestion
Each microservice has a service user whose roles are defined in the `cumulocity.json` manifest. By default, the platform API services provided by the Microservice SDK operate in the tenant scope. The following example uses the predefined eventApi bean in tenant scope:
```

> Each microservice has a service user whose roles are defined in the `cumulocity.json` manifest. By default, the Platform API services provided by the Microservice SDK operate in the tenant scope. The following example uses the predefined eventApi bean in tenant scope:

Reviewer: ```suggestion
Each microservice has a service user whose roles are defined in the `cumulocity.json` manifest. By default, the platform API services provided by the Microservice SDK operate in the tenant scope. The following example uses the predefined eventApi bean in tenant scope:
```

| alarmApi, tenantAlarmApi                                       | userAlarmApi                    |
| eventApi, tenantEventApi                                       | userEventApi                    |
| eventBinaryApi, tenantEventBinaryApi                           | userEventBinaryApi              |
| auditRecordApi, tenantAuditRecordApi                           | userAuditRecordApi              |
| deviceCredentialsApi, tenantDeviceCredentialsApi               | userDeviceCredentialsApi        |
| binariesApi, tenantBinariesApi                                 | userBinariesApi                 |
| userApi, tenantUserApi                                         | userUserApi                     |
| tenantOptionApi, tenantTenantOptionApi                         | userTenantOptionApi             |
| systemOptionApi, tenantSystemOptionApi                         | userSystemOptionApi             |
| tokenApi, tenantTokenApi                                       | userTokenApi                    |
| notificationSubscriptionApi, tenantNotificationSubscriptionApi | userNotificationSubscriptionApi |

> Various examples demonstrating use cases of the Platform API can be found in the GitHub repositories

Reviewer: ```suggestion
Various examples demonstrating use cases of the platform API can be found in the GitHub repositories
```

---

> The Cumulocity Microservice SDK is a tool kit supporting efficient development of applications

Reviewer: ```suggestion
The {{< product-c8y-iot >}} Microservice SDK is a tool kit supporting efficient development of applications
```

> based on the Cumulocity Platform API. The source code is located in the Github repository

Reviewer: ```suggestion
based on the {{< product-c8y-iot >}} Platform API. The source code is located in the GitHub repository
```

@EnableMicroservicePlatformInternalApi | Injects the platform API services into Spring context for a microservice to use
@EnableTenantOptionSettings | Provides microservice configuration within tenant options and allows overriding default properties from files

### Platform API {#platform-api}

> The Cumulocity Microservice SDK package consists of a number of services that are built and injected into Spring context.

Reviewer: ```suggestion
The {{< product-c8y-iot >}} Microservice SDK package consists of a number of services that are built and injected into the Spring context.
```

> A developer can use them to perform basic operations against the platform. The beans are built based on the properties read from a file. For hosted deployment, most of the properties are provided by the platform.

Reviewer: ```suggestion
A developer can use them to perform basic operations against the platform. The beans are built based on the properties read from a file. For hosted deployments, most of the properties are provided by the platform.
```

The API provides the following services:

* Alarm - AlarmApi
* AuditRecord - AuditRecordApi
* Operation - DeviceControlApi
* Event - EventApi
* EventBinary - EventBinaryApi
* ExternalID - IdentityApi
* Binary - BinariesApi
* ManagedObject - InventoryApi
* Measurement - MeasurementApi
* DeviceCredentials - DeviceCredentialsApi
* User - UserApi
* TenantOption - TenantOptionApi
* SystemOption - SystemOptionApi
* Token - TokenApi
* Notification - NotificationSubscriptionApi

The API supports basic CRUD methods. The following is an alarm interface example:

```java
// Methods
AlarmRepresentation create(final AlarmRepresentation alarm)
Future createAsync(final AlarmRepresentation alarm)

AlarmRepresentation getAlarm(final GId gid)
AlarmCollection getAlarms()
AlarmCollection getAlarmsByFilter(final AlarmFilter filter)

AlarmRepresentation update(final AlarmRepresentation alarm)
```

> Sample usage:

Reviewer 3: If new users will follow the documentation and copy & paste this it will most probably result in a "not within any context" exception.
We should make clear that each bean must be executed in a context already e.g. adding a comment //Called from within a context see chapter....

Author: The SDK provides the API beans in the tenant scope by default by attaching the @Primary annotation in the bean definition; a qualifier is not required:

    @Primary
    @TenantScope
    @Bean(name = {"eventApi", "tenantEventApi"})
    public EventApi getEventApi() throws SDKException {
        return delegate.getEventApi();
    }
		
An additional example explains the equivalence of

    @Autowired
    private EventApi eventApi;

and

    @Autowired
    @Qualifier("tenantEventApi")
    private EventApi eventApi;

Reviewer 3: It's not about the tenant scope or any qualifier, it's about being in any context which is not guided and part of this example (but should be).

Reviewer 3: Please try out the given example in a blank spring project e.g. App.java and you will recognized that you get an exception that way.

```java
@Autowired
private AlarmApi alarms;

public AlarmRepresentation addHelloAlarm (){
    AlarmRepresentation alarm = new AlarmRepresentation();
    alarm.setSeverity("CRITICAL");
    alarm.setStatus("Hello");

    return alarms.create(alarm);
}
```

### Using the Platform API {#using-platform-api}

> The Cumulocity Microservice SDK provides predefined methods to interact with the Cumulocity platform.

Reviewer: ```suggestion
The {{< product-c8y-iot >}} Microservice SDK provides predefined methods to interact with the {{< product-c8y-iot >}} platform.
```

> Several APIs are available to execute basic operations on the platform with corresponding Beans in the SDK,

Reviewer: ```suggestion
Several APIs are available to execute basic operations on the platform with corresponding beans in the SDK,
```

> like `eventApi`, `inventoryApi`, `measurementApi`, etc.

Reviewer: ```suggestion
such as `eventApi`, `inventoryApi`, `measurementApi`.
```

> There are two types of context in which API requests can be executed: tenant scope and user scope.

Reviewer 3: Here we should add a more highlighted box: "Every API Bean must be executed in a context". Here are the options to get in that context: ..."

Author: Text now in bold print

Each microservice has a service user which can be used for the interaction with the platform.
The roles associated with this user are specified in the manifest.
Within the tenant scope, the credentials of this service user are used for the communication
with the platform, while within the user scope the credentials of the authenticated user
sending the request to the microservice are used. By default, the Platform API related beans provided by the Microservice SDK are created in the
tenant scope.

A basic example shows the usage of the predefined `eventApi` in the tenant scope:

```java
@Autowired
private EventApi eventApi;

> public List<EventRepresentation> getEvents () {

Reviewer 3: Same comment as above. Explaining that there are types of context but no example how to get in that context... --> Will result in a "not within any context" exception.

Author: please see first comment above

  return eventApi.getEvents().get().getEvents();
}
```

If the request has to be executed within the user scope the bean must be annotated with  
`@Qualifier("userEventApi")` like in the next example.

```java
@Autowired
@Qualifier("userEventApi")
private EventApi userEventApi;

public List<EventRepresentation> getEvents () {
  return userEventApi.getEvents().get().getEvents();
}
```

> Default context is the tenant scope also in Java classes with `@RestController` annotations

Reviewer: ```suggestion
The default context is the tenant scope also in Java classes with `@RestController` annotations
```

> The default context is the tenant scope also in Java classes with `@RestController` annotations

Reviewer 3: One of the most important sentence in this chapter! Please highlight like > **Important**

Author: I'm not sure about this. I could not find any code tying @RestController or @EventListener to a context.
Isn't it like this:
The context is explicitly set by wiring the desired API bean or by the MicroserviceSubscriptionsService or ContextService wherever a context is needed.

Reviewer 3: No, it's not working like that. You will run into "not within any context" if you just wire the beans without any context. The exceptions are the eventlistener and controllers... where you don't need to use the subscriptionsService because you already in the tenant context. Please test it yourself.

and in `EventListener` methods.
To access a specific tenant, the appropriate context needs to be set by using the
`MicroserviceSubscriptionsService` like in the next example.

```java
@Autowired

> private MicroserviceSubscriptionsService subscriptionsService;

Reviewer 3: This is actually the first full working example for the reader without the probability to run in the "not within any context" exception.

Author: The qualifier is required here because the "userEventAPI" is not annotated with @Primary

void deleteAlarmsByFilter(AlarmFilter filter) throws IllegalArgumentException, SDKException;
```

> This is an example of retrieving all devices registered in the `management` tenant:

Reviewer 3: .. in all subscribed tenants.

Author: text adapted accordingly

```java
@Autowired
MicroserviceSubscriptionsService subscriptionsService;

@Autowired
InventoryApi inventoryApi;

public List<ManagedObjectRepresentation> getAllDevicesTenant2() {
  List<ManagedObjectRepresentation> managedObjectsList = new ArrayList<>();

>   subscriptionsService.runForTenant("management", () -> {

Reviewer 3: Better to replace with `subscriptionsService.runForEachTenant()`.
Management is for 2 reasons wrong here:
1. It should be the tenantID which in most customer cases starts with t<number> not the tenant name.
2. In SaaS environments users don't have access to any management tenant. So it doesn't exist for them.

Author: Now using `runForEachTenant` and `inventoryApi.getManagedObjectsByFilter(filter)`

$ mvn microservice:upload -Dupload.application.name=helloworld -Dupload.url=https://demos.cumulocity.com -Dupload.username=demos/username -Dupload.password=****** -Dskip.microservice.upload=false
```

#### Heap and perm/metadata {#heap-and-permmetadata}

> To calculate heap and perm/metadata, the SDK takes the limit defined in the [microservice manifest](/microservice-sdk/general-aspects/#microservice-manifest) and converts it into Megabytes (MB). The minimal value is 178MB. 10% is reserved for "system", but not less than 50 MB. 10% is taken for Metaspace, but not less than 64 MB and not more than 1024MB. The rest is allocated for heap size.

Reviewer: ```suggestion
To calculate heap and perm/metadata, the SDK takes the limit defined in the [microservice manifest](/microservice-sdk/general-aspects/#microservice-manifest) and converts it into megabytes (MB). The minimal value is 178MB. 10% is reserved for "system", but not less than 50 MB. 10% is taken for metaspace, but not less than 64 MB and not more than 1024MB. The rest is allocated for heap size.
```
